daemon/api_sideload_n_try.go

```diff
@@ -207,6 +205,118 @@ out:
 	return AsyncResponse(nil, chg.ID())
 }
 
+// maxReadBuflen is the maximum buffer size for reading the non-file parts in the snap upload form
+const maxReadBuflen = 1024 * 1024
+
+// readForm returns a Form populated with values (for non-file parts) and file headers (for file
+// parts). The file headers contain the original file name and a path to the persisted file in
+// dirs.SnapDirBlob. If an error occurs and a non-nil Response is returned, an attempt is made
+// to remove temp files.
+func readForm(reader *multipart.Reader) (_ *Form, apiErr *apiError) {
+	maxMemory := int64(maxReadBuflen)
+	form := &Form{
+		Values:   make(map[string][]string),
+		FileRefs: make(map[string][]*FileReference),
+	}
+
+	// clean up if we're failing the request
+	defer func() {
+		if apiErr != nil {
+			form.RemoveAll()
+		}
+	}()
+
+	for {
+		part, err := reader.NextPart()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, BadRequest("cannot read POST form: %v", err)
+		}
+
+		name := part.FormName()
+		if name == "" {
+			continue
+		}
+
+		filename := part.FileName()
+		if filename == "" {
+			// non-file parts are kept in memory
+			buf := &bytes.Buffer{}
+
+			// copy one byte more than the max so we know if it exceeds the limit
+			n, err := io.CopyN(buf, part, maxMemory+1)
+			if err != nil && !errors.Is(err, io.EOF) {
+				return nil, BadRequest("cannot read form data: %v", err)
+			}
+
+			maxMemory -= n
+			if maxMemory < 0 {
+				return nil, BadRequest("cannot read form data: exceeds memory limit")
+			}
+
+			form.Values[name] = append(form.Values[name], buf.String())
+			continue
+		}
+
+		tmpPath, err := writeToTempFile(part)
+
+		// add it to the form even if err != nil, so it gets deleted
+		ref := &FileReference{TmpPath: tmpPath, Filename: filename}
+		form.FileRefs[name] = append(form.FileRefs[name], ref)
+
+		if err != nil {
+			return nil, InternalError(err.Error())
+		}
+	}
+
+	// sync the parent directory where the files were written to
+	if len(form.FileRefs) > 0 {
+		dir, err := os.Open(dirs.SnapBlobDir)
+		if err != nil {
+			return nil, InternalError("cannot open parent dir of temp files: %v", err)
+		}
+		defer func() {
+			if cerr := dir.Close(); err == nil && cerr != nil {
```

Comment:
Is err == nil check here achieving anything? We know err is nil after os.Open (otherwise we would have returned InternalError above), or am I confused?

Reply:
No, you're right, `err` should be `apiErr`

Reply:
Fixed, thanks!